query/src/sessions/session_mgr.rs

```diff
@@ -132,34 +124,46 @@ impl SessionManager {
         self.storage_cache_manager.as_ref()
     }
 
-    pub fn create_session(self: &Arc<Self>, typ: impl Into<String>) -> Result<SessionRef> {
-        let mut sessions = self.active_sessions.write();
-        match sessions.len() == self.max_sessions {
-            true => Err(ErrorCode::TooManyUserConnections(
-                "The current accept connection has exceeded mysql_handler_thread_num config",
-            )),
-            false => {
-                let session = Session::try_create(
-                    self.conf.clone(),
-                    uuid::Uuid::new_v4().to_string(),
-                    typ.into(),
-                    self.clone(),
-                )?;
+    pub async fn create_session(self: &Arc<Self>, typ: impl Into<String>) -> Result<SessionRef> {
+        {
+            let sessions = self.active_sessions.read().await;
+            if sessions.len() == self.max_sessions {
+                return Err(ErrorCode::TooManyUserConnections(
+                    "The current accept connection has exceeded mysql_handler_thread_num config",
+                ));
+            }
+        }
+        let session = Session::try_create(
+            self.conf.clone(),
+            uuid::Uuid::new_v4().to_string(),
+            typ.into(),
+            self.clone(),
+        )
+        .await?;
 
-                label_counter(
-                    super::metrics::METRIC_SESSION_CONNECT_NUMBERS,
-                    &self.conf.query.tenant_id,
-                    &self.conf.query.cluster_id,
-                );
+        label_counter(
+            super::metrics::METRIC_SESSION_CONNECT_NUMBERS,
+            &self.conf.query.tenant_id,
+            &self.conf.query.cluster_id,
+        );
+        let mut sessions = self.active_sessions.write().await;
+        if sessions.len() < self.max_sessions {
+            sessions.insert(session.get_id(), session.clone());
 
-                sessions.insert(session.get_id(), session.clone());
-                Ok(SessionRef::create(session))
-            }
+            Ok(SessionRef::create(session))
```

Comment:
The metrics counter should be moved into the if-closure.

But I am not sure if `label_counter` contains any async calls in it.